### ISSUE TITLE
ofi/psm2: capability mismatch between user's library and hard-coded values

### ISSUE THREAD

**Issue by danghvu** [pmodels/mpich#2952]
**Summary of the issue:**

* MPICH uses hard-coded value for PSM2 capability in `ofi_capability_sets.h` which may not be correct in users' installation (e.g. scalable end-point feature).
* There is no detection at the MPICH layer to realize such mismatch.
* `libfabric` can detect and install with lower capability but this information is not forwarded.

**Some proposed solutions:** 

* "Update the version of libfabric embedded in MPICH and to replicate whatever check the psm2 provider is doing against the psm2 library" [wesbland]
* "MPICH detects psm2 doesn't support scalable endpoint, it would modify the capability set variable for psm2 so it wouldn't require scalable endpoint" [hajimefu]
* "check at runtime to make sure these capability sets are indeed provided and abort if they are not (with a helpful error message.." [pavan]
* "multiple static capability sets, e.g., for psm2, we might need one that forces scalable endpoints and one that does not" [pavan]
* Use an embedded psm2 if "ofi:psm2" is specified [danghvu] 

Original issue:
I got an `Abort (15)` during `MPI_Init` after sync up to master on a KNL cluster (bbop knlall) using 57e75e9.
However, if I use 3711bf1d (one behind) it works as usual. 

Configuration:
```
../configure --prefix=/home/danghvu/ch4-master --with-device=ch4:ofi:psm2 --enable-fast=all,O3 --disable-perftest --disable-ft-tests --with-fwrapname=mpigf --with-filesystem=ufs+nfs --enable-timer-type=linux86_cycle --enable-romio --with-mpe=no --with-smpcoll=yes --with-assert-level=0 --enable-shared --enable-static --enable-error-messages=yes --enable-visibility --enable-large-tests --enable-g=none --enable-timing=none --enable-error-checking=no --enable-fast=all,O3 --disable-debuginfo --with-device=ch4:ofi:psm2 --enable-handle-allocation=default --enable-threads=multiple --without-valgrind --enable-timing=none --enable-ch4-shm=yes --enable-thread-cs=global --with-ch4-netmod-ofi-args= MPICHLIB_CFLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic -std=gnu99 MPICHLIB_CXXFLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic MPICHLIB_FCFLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic MPICHLIB_F77FLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic MPICHLIB_LDFLAGS=-O3 -L/usr/lib64  -msse2 -msse4.2 -mcrc32 -mavx512f -mtune=generic
```

**Comment by wesbland** [comment]
Do you have the latest version of the PSM2 library? @hajimefu can give you more information, but you need to have a recent enough version of both libfabric and PSM2 to use scalable endpoints.

**Comment by pavanbalaji** [comment]
I hope the latest `libfabric` was already embedded into `mpich`, when the scalable endpoints patch went in.  We only need to worry about the `psm2` version, but I hope we are catching such problems at configure time?

**Comment by wesbland** [comment]
For all the reasons that came up during the submodules work, I really don't want to keep updating the embedded libfabric version all of the time (we actually don't ever use embedded libfabric anymore).

I don't know exactly how your setup looks anymore, but I was under the impression that for regular testing, you guys were still using an external libfabric library.

**Comment by raffenet** [comment]
We test both now. Embedded libfabric is version 1.5.1, which should contain the support for scalable endpoints with psm2.

**Comment by wesbland** [comment]
Yep. That version should be fine. Glad to hear you're testing embedded builds.

**Comment by pavanbalaji** [comment]
I don't think we should keep updating the embedded `libfabric` unless we are missing a feature, or we want to move to a new release of `libfabric`.  That's the same as what we do with `hwloc`.

**Comment by wesbland** [comment]
I agree with that, though lately we've been leaning hard on latest and greatest code, particularly with respect to PSM2.

**Comment by pavanbalaji** [comment]
That's fine to do using external `libfabric` builds.  I was referring to updating the version of the embedded `libfabric` submodule.

**Comment by danghvu** [comment]
I have tried a few things:
* I am able to run with latest `libfabric` and `psm2` (have to install myself and use `--with-libfabric` and additionally add `LDFLAGS="-L$MY_PSM2_LIB_DIR -lpsm2"` to `configure`).
* I am able to run with the embedded libfabric by adding additionally `CFLAGS="-I$MY_PSM2_INCLUDE_DIR"`
* I am _unable_ to run with the embedded libfabric when using `LD_PRELOAD` or `LD_LIBRABRY_PATH` with my psm2 installation which is weird since psm2 is loaded dynamically (I made sure ldd show the correct one). Is the capability of psm2 determined at compiled time ?

I concur with Pavan that this capacity mismatch should be detected at configuration time if I choose to use the embedded libfabric. It would also be nice to be able to pass `--with-psm2=...` to the `configure` of MPICH if psm2 can't be reloaded dynamically.

**Comment by pavanbalaji** [comment]
This was working earlier.  Which commit broke it?  Do we know?

**Comment by hajimefu** [comment]
@danghvu 
> Is the capability of psm2 determined at compiled time ?
I think so, I mean, IIRC libfabric looks into psm2 at the configure time to see if scalable endpoints (=multiple endpoints) are available. cf. https://github.com/ofiwg/libfabric/pull/3705
Have you tried `--enable-psm2=$psm2_dir` when configuring MPICH for embedding OFI?

@pavanbalaji As @danghvu mentioned I think it's 57e75e9.

**Comment by pavanbalaji** [comment]
We should be more careful when adding these patches in.  Didn't we look for jenkins tests?

**Comment by raffenet** [comment]
We don't test `psm2` in our Jenkins. Regardless if `libfabric` can detect support for scalable endpoints, it's hard-coded to enabled in `ch4/ofi/psm2` configuration.

**Comment by pavanbalaji** [comment]
I see.  That sounds like a `libfabric` bug then.

**Comment by raffenet** [comment]
@hajimefu can `ofi_init.h` gracefully handle when support is not available? Do we just revert this patch for now until we figure out a better solution?

**Comment by wesbland** [comment]
I'm not sure what the best solution is here. The reason we have those capability sets is to create a fast path to remove all of the branches for known configurations. Obviously with versions that are older than we expect, those assumptions might be broken (as in this case).

We assume that you have an up to date version of both of those libraries. If that's not true, then using the runtime capability set should do all of the proper detection (change `--with-device=ch4:ofi:psm2` to `--with-device=ch4:ofi`).

**Comment by wesbland** [comment]
The alternative is to have a very fancy configure-time detection for all of the capabilities in `ofi_capability_sets.h`.  I'm not sure if that's actually possible in all cases or not (haven't looked to see if everything is portably detectable).

**Comment by pavanbalaji** [comment]
@wesbland I think we are overcomplicating things here.  We already embed `libfabric` in mpich.  Is there a bug in that version?  If yes, then we need to simply upgrade that version.  If the bug is in `psm2`, then `libfabric` needs to check at configure time for >= some version of psm2.

**Comment by wesbland** [comment]
Oh ok. That's a much easier solution. 😄 Let's do that. Check for whatever the most recent version of psm2 if you use a compile-time psm2 capability set.

**Comment by pavanbalaji** [comment]
To be clear, I prefaced that with a question: `is there a bug in that version`?  I'm not recommending blindly requiring the latest version of `psm2`, which would be silly.

**Comment by pavanbalaji** [comment]
Where is the bug exactly?  Is `psm2` exposing the wrong capability?  Is `libfabric` not detecting the capability correctly?

**Comment by pavanbalaji** [comment]
@wesbland To be clear, you are referring to the hard-coded capability sets in `mpich`, correct?  Not in `libfabric`.  In that case, we can force `mpich` to pick those hard-coded capability sets only when a minimum version of `libfabric` is used.  Is there a way to force it to use a minimum version of `psm2` given that it's hidden behind the `libfabric` library?

**Comment by wesbland** [comment]
Correct. I'm referring to the MPICH code.

I was just talking to Hajime about this and I think we realized what the problem is. Inside libfabric, when it detects the installed psm2 library, it seems to automatically detect that what it needs for scalable endpoints isn't supported and therefore turns it off. We (MPICH) don't realize that and still blindly require scalable endpoints, which the psm2 provider no longer supports (because it was compiled out).

So one solution would be to update the version of libfabric embedded in MPICH and to replicate whatever check the psm2 provider is doing against the psm2 library to check for scalable endpoint support. If that check passes, we can fail the configure and say you need an updated psm2 library (or we could fall back to the runtime capability set if we didn't want to be so harsh, but the performance degradation might be a surprise).

**Comment by hajimefu** [comment]
Another solution is similar to what @wesbland described above, but if MPICH knows psm2 doesn't support scalable endpoint, it would modify the capability set variable for psm2 so it wouldn't require scalable endpoint (basically, going back to the state before 57e75e9.) This way we don't have to fall back to runtime capability sets.

**Comment by pavanbalaji** [comment]
`mpich` does not have access to `psm2`.  It only has access to `libfabric` and replicating that logic instead `mpich` seems gross and is a maintenance hassle.

Besides, I think that this is a bigger problem than just `psm2` and scalable endpoints.  For any given combination of features and providers, the static hardcoding is essentially relying on user knowledge about what the provider gives.  If the "user knowledge" (in this case, the version of `libfabric` and `psm2` that I'm using have support scalable endpoints) is incorrect, then that's a user error.

Here are a few things that we could do:

1. It's OK to statically let the user pick capability sets (e.g., on a large machine, where the system administrator controls all libraries), but we should check at runtime to make sure these capability sets are indeed provided and abort if they are not (with a helpful error message to the user saying that they picked 'xxx' feature at configure time, but the provider doesn't actually seem to have this feature, so go fix your configure line).  This happens at init time and should not add any overhead to the direct branch-free path.

2. We need to have multiple static capability sets, e.g., for `psm2`, we might need one that forces scalable endpoints and one that does not.  So, if the `psm2` library that I have does not support scalable endpoints, I can still build `mpich` with a static capability set, but one that disables scalable endpoints.  Say `PSM2_SCALABLE_ENDPOINTS` and `PSM2_NOSCALABLE_ENDPOINTS`.  I admit that this model will lead to a significant increase in capability sets if we create all combinations.  But the intent of these static capability sets is to figure out what large machines indeed need, and only provide those, and everyone else goes through the `runtime` option.  So, for example, if no large machine needs `PSM2_NOSCALABLE_ENDPOINTS`, then we don't need to provide that as a static capability set.

**Comment by danghvu** [comment]
As a user, I would expect some meaningful error messages at the least, preferably at configuration time / compile time (as there is no way to fix it at runtime). 

A suggestion/question, if the capability is hard-coded, could MPICH just maintains a submodule of psm2 (with appropriate commit for such capability) and compile it if I choose `ch4:ofi:psm2`? This is already done for a few other things in MPICH.

@hajimefu 
> Have you tried --enable-psm2=$psm2_dir when configuring MPICH for embedding OFI?

I remember have tried, it compiled but still was not working -- I can try again, but I do not see this flag in `./configure --help` of MPICH so I suspect I should use it.

**Comment by wesbland** [comment]
@pavanbalaji - I agree with that idea. I believe at least some part of this already exists in `ofi_init.h`, but obviously it must not be complete enough to catch this problem. Can we create an issue that is more descriptive for that so someone can work on it when they have some time (or even just repurpose this issue by updating the description/title)?

@danghvu - Do you mean a submodule of the entire psm2 library? I think at that point, we're adding far more internal dependencies than we want, especially considering there will be plenty of times where MPICH would not need PSM2. I think that's different than hwloc/libfabric/ucx where it's highly likely that you're going to be using those things (even if you won't use libfabric and ucx at the same time).

Also, I believe if you're building with an embedded libfabric and you pass `--enable-psm2` to the MPICH configure, it will get passed down to libfabric (along with any other configure flags). I assume the flag you want is `--with-psm2=<something>` though as I assume that `--enable-psm2` tells libfabric to enable the psm2 provider.

**Comment by pavanbalaji** [comment]
@wesbland it might be better to repurpose this issue, since it already contains a lot of the discussion as to how we got to this point.

Btw, I'm not sure `--enable-psm2` does anything to what `libfabric` picks apart from setting the `CFLAGS` and other flags.  For example, if the `libfabric` library is built statically with some `psm2` library, this option will have no impact.

**Comment by wesbland** [comment]
@pavanbalaji That sounds good.

@danghvu - Can you summarize this conversation in the description/title above?

**Comment by hajimefu** [comment]
@danghvu As a side discussion, it looks like libfabric does support embedding psm2 in it (https://github.com/ofiwg/libfabric/pull/3612), but as @wesbland mentioned I'm reluctant to this solution, as it'll add more external dependencies to MPICH. For example, in order to build psm2 you need some header files from HFI drivers, which are typically not available in many systems.

**Comment by hzhou** [comment]
PR #5233 now will adjust global settings based on the final provider info returned from `fi_getinfo` (with hints set from the `ofi_capability_sets.h`). If the runtime libfabric return less capability, the global settings will be adjusted down. In particular, reference https://github.com/hzhou/mpich/blob/bc5a69e3a47d7d62ea64dc3e254bf616acf19ec5/src/mpid/ch4/netmod/ofi/init_settings.c#L326-L328. Thus, I believe the described issue https://github.com/pmodels/mpich/issues/2952#issue-292222245 no longer exist.
